expression/builtin_compare_vec.go

```diff
@@ -214,11 +214,56 @@ func (b *builtinLeastRealSig) vecEvalReal(input *chunk.Chunk, result *chunk.Colu
 }
 
 func (b *builtinLeastStringSig) vectorized() bool {
-	return false
+	return true
 }
 
 func (b *builtinLeastStringSig) vecEvalString(input *chunk.Chunk, result *chunk.Column) error {
-	return errors.Errorf("not implemented")
+	if err := b.args[0].VecEvalString(b.ctx, input, result); err != nil {
+		return err
+	}
+	if len(b.args) == 1 {
+		return nil
+	}
+
+	n := input.NumRows()
+	arg, err := b.bufAllocator.get(types.ETString, n)
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(arg)
+
+	dst, err := b.bufAllocator.get(types.ETString, n)
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(dst)
+
+	src := result
+	for j := 1; j < len(b.args); j++ {
+		if err := b.args[j].VecEvalString(b.ctx, input, arg); err != nil {
+			return err
+		}
+
+		src.MergeNulls(arg)
+		for i := 0; i < n; i++ {
+			if src.IsNull(i) {
+				dst.AppendNull()
+				continue
+			}
+			srcStr := src.GetString(i)
+			argStr := arg.GetString(i)
+			if types.CompareString(srcStr, argStr) < 0 {
+				dst.AppendString(srcStr)
+			} else {
+				dst.AppendString(argStr)
+			}
+		}
+		src, dst = dst, src
```

Comment:
When the `len(b.args)` is even, the loop will run odd times, and `dst` will point to `result`.
Because of L239,  it will be collected by `bufAllocator` when the function return.  So the `result` maybe realloc and be used for other vec_functions. It will cause data race. 
I think you need to adjust your code.

Reply:
I will fix it.

Reply:
And please update the bench result to compare the performance for the new implement. thanks.

Reply:
ok

Reply:
go test -v -benchmem -bench=BenchmarkVectorizedBuiltinCompareFunc -run=BenchmarkVectorizedBuiltinCompareFunc -args 'builtinLeastStringSig'
goos: linux
goarch: amd64
pkg: github.com/pingcap/tidb/expression
BenchmarkVectorizedBuiltinCompareFunc/builtinLeastStringSig-VecBuiltinFunc-4         	   58890	     19649 ns/op	       0 B/op	       0 allocs/op
BenchmarkVectorizedBuiltinCompareFunc/builtinLeastStringSig-NonVecBuiltinFunc-4      	   13320	     90633 ns/op	       0 B/op	       0 allocs/op
BenchmarkVectorizedBuiltinCompareFunc/builtinLeastStringSig-VecBuiltinFunc#01-4      	   29248	     41915 ns/op	       0 B/op	       0 allocs/op
BenchmarkVectorizedBuiltinCompareFunc/builtinLeastStringSig-NonVecBuiltinFunc#01-4   	   10000	    116259 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/pingcap/tidb/expression	6.344s

Reply:
the performance compare for the new implement
```
go test -v -benchmem -bench=BenchmarkVectorizedBuiltinCompareFunc -run=BenchmarkVectorizedBuiltinCompareFunc -args 'builtinLeastStringSig'
goos: linux
goarch: amd64
pkg: github.com/pingcap/tidb/expression
BenchmarkVectorizedBuiltinCompareFunc/builtinLeastStringSig-VecBuiltinFunc-4         	   58890	     19649 ns/op	       0 B/op	       0 allocs/op
BenchmarkVectorizedBuiltinCompareFunc/builtinLeastStringSig-NonVecBuiltinFunc-4      	   13320	     90633 ns/op	       0 B/op	       0 allocs/op
BenchmarkVectorizedBuiltinCompareFunc/builtinLeastStringSig-VecBuiltinFunc#01-4      	   29248	     41915 ns/op	       0 B/op	       0 allocs/op
BenchmarkVectorizedBuiltinCompareFunc/builtinLeastStringSig-NonVecBuiltinFunc#01-4   	   10000	    116259 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/pingcap/tidb/expression	6.344s
```